[BOLT] Improve DWARF CFI generation for pac-ret binaries

bgergely0 opened the pull request:
During InsertNegateRAState pass we check the annotations on instructions,
to decide where to generate the OpNegateRAState CFIs in the output binary.

As only instructions in the input binary were annotated, we have to make
a judgement on instructions generated by other BOLT passes.
Incorrect placement may cause issues when an (async) unwind request
is received during the new "unknown" instructions.

This patch adds more logic to make a more informed decision on by taking into account:
- unknown instructions in a BasicBlock with other instruction have the
  same RAState. Previously, if the BasicBlock started with an unknown instruction,
  the RAState was copied from the preceding block. Now, the RAState is copied from
  the succeeding instructions in the same block.
- Some BasicBlocks may only contain instructions with unknown RAState,
  As explained in issue #160989, these blocks already have incorrect unwind 
  info. Because of this, the last known RAState based on the layout order is copied.

Updated bolt/docs/PacRetDesign.md to reflect changes.

bgergely0 committed:
[BOLT] Improve InsertNegateRAStatePass::inferUnknownStates

Previous implementation used a simple heuristic. This can be improved in
several ways:
- If a BasicBlock has instruction both with known RAState and unknown RAState,
  use the known states to work out the unknown ones.
- If a BasicBlock only consists of instructions with unknown RAState,
  use the last known RAState from its predecessors, or the first known
  from its successors to set the RAStates in the BasicBlock. This includes
  error checking: all predecessors/successors should have the same RAState.
- Some BasicBlocks may only contain instructions with unknown RAState,
  and have no CFG neighbors. These already have incorrect unwind info.
  For these, we copy the last known RAState based on the layout order.

Updated bolt/docs/PacRetDesign.md to reflect changes.

bgergely0 commented:
## How to test this?

I have a few concern on testing:
- the patch adds several "helper functions", which would all ideally be unittested, but as they are private functions of the pass, I cannot "export" them to be visible for tests.
- High-level lit testing is also not feasible. The change tries to cover many low-probability edge-cases, all in the scope of "newly generated instructions",  that is instructions generated by other BOLT passes. This makes creating lit tests difficult, as I would have to create an input assembly on which the other passes would do the exact transformations that would trigger/exercise these newly added functions. Future changes to any other pass (running before InsertNegateRAStates) could also break these tests.

### Proposal
I think the following plan is more reasonable:
- create "single-pass" unittests in gtest, similar to the MCPlusBuilder unittests. This way I can manually create the exact BinaryFunctions with the exact RAState annotations that exercise additions to this patch. Then, printing the BinaryFunction after the pass runs would be enough to check the locations of newly inserted CFIs.

I also think this single-pass gtest way of testing would be useful in other future additions to BOLT.

bgergely0 committed:
[BOLT] Single-pass unittest for InsertNegateRAState

This commit creates a new directory: bolt/unittests/Passes, to be used
by unittests that need to register and run passes with the
BinaryFunctionPassManager.

An example test is created for InsertNegateRAState pass. Actual tests
will be added in followup commits.

bgergely0 committed:
[BOLT] Test fillUnknownRAStateInBB

bgergely0 committed:
[BOLT] Test fillUnknownStubs

bgergely0 commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -91,44 +105,214 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
-          RAState = PrevRAState ? *PrevRAState : false;
+          if (!PrevRAState) {
+            llvm_unreachable(
+                "Previous Instruction has no RAState in fillUnknownStubs.");
```
TODO: change to fatal BOLT error, based on the discussion in #162820

paschalis-mpeis commented on bolt/docs/PacRetDesign.md:
```diff
@@ -200,16 +200,29 @@ This pass runs after optimizations. It performns the _inverse_ of MarkRAState pa
+
+3. If a BasicBlock has no CFG neighbors, we have to copy the RAState of the previous
+    BasicBlock in layout order.
+
```
This must be for the case that LongJmp creates stubs for a call specifically, ie [here](https://github.com/llvm/llvm-project/blob/224a7176dc6afaec83bfcfdb7542d30a130997cc/bolt/lib/Passes/LongJmp.cpp#L257) right?

bgergely0 replied:
exactly.

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -91,44 +105,214 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
+    return;
+  // If the first instruction has unknown RAState, we should copy the first
+  // known RAState.
+  auto RAStateOpt = BC.MIB->getRAState(*First);
```
This is for my own understanding:

In this case, is it okay to copy from the first known state, which follows the first unknown instruction? Is this a case that occurs in practice?

bgergely0 replied:
I don't think I encountered it in practice. I included it because:
- I don't have a "global view" of all passes, with each insertion exactly understood,
- even if it does not occur currently, it could happen later, as BOLT is changing,
- if we don't assign an RA state to _each and every_ instruction here, the pass will crash later, as we expect all to have an RA state

So we have to add an RA state, and this is a more sophisticated method, then copying the RA state of the previous BB.

As of soundness: all instructions in a BB have the same RA state, except:
- if the BB has psign/pauth
- if a BB has a noreturn call, and it is incorrectly parsed as one BB instead of two (see: https://github.com/llvm/llvm-project/issues/115154)
I do not think BOLT ever inserts a new call to a noreturn function, so this should be correct.

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -91,44 +105,214 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
+  // At this point we know the RAState of the first instruction,
+  // so we can propagate the RAStates to all subsequent unknown instructions.
+  MCInst Prev = *First;
+  for (auto It = BB.begin() + 1; It != BB.end(); ++It) {
```
Any reason not to start at Prev+1 ?

bgergely0 replied:
As a first glance, that should also be correct. Although it wouldn't be a big difference, as we skip CFIs early in the loop.

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -91,44 +105,214 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
+
+    auto PrevRAState = BC.MIB->getRAState(Prev);
+    if (!PrevRAState)
+      llvm_unreachable("Previous Instruction has no RAState.");
```
The only gaps that we are supposed to fill are on PSign/PAuth instructions?

Since we enter this loop after finding an initial state (ie on *First), we could do this check on the current iteration, instead of the previous potentially unknown state?
For example:

```
if !RAState {
  if isPSign -> true
  elif isPAuth -> false
  else -> fatal()
}
```

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -91,44 +105,214 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
+      MCInst &LastInst = *LI;
+      checkRAState(NeighborRAState, LastInst);
+    }
+  } else if (BB.succ_size() != 0) {
```
Does this need to be an `else if`, or shall we confirm a matching state with successors as well?

paschalis-mpeis commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -91,44 +105,214 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
+    std::pair<int, bool> Iter = fillUnknowns(BF);
+    if (Iter.first == 0)
+      break;
+    if (!Iter.second) {
```
This happens when none of the sucs/preds were able to provide us with a state, right? Would it be more accurate to say "ignored" instead of "not optimized"?

bgergely0 replied:
I wrote it like this, because I noticed other passes also say "will not be optimized" to users, when a function is ignored.
Example: https://github.com/llvm/llvm-project/blob/main/bolt/lib/Passes/PatchEntries.cpp#L102-L106
I think of the two wordings as equivalent, just thought I'd stick to what is used elsewhere. Might also need to reword MarkRAStates' message about ignored functions...

paschalis-mpeis reviewed (commented):
Hey Gergely,

Thanks for your patch. This is not a full review of the patch.
I've added a few comments that may benefit from a bit more context, and some points for clarification.

bgergely0 committed:
[BOLT] Review changes

maksfb commented:
Hi Gergely, thanks for the patch. For testing purposes, if we had a way to run a selected optimization pass without running anything else, would it help? E.g. `--pass-list=mark-ra-states,insert-negate-ra-state-pass,finalize-functions` will only run those three and not the rest. Some passes that are run at the end, e.g. `InstructionLowering` are mandatory and we can add them by default. But you will have  the ability to test IR after every pass using `--print-after...`.

bgergely0 commented:
Hi @maksfb! 

> For testing purposes, if we had a way to run a selected optimization pass without running anything else, would it help? E.g. --pass-list=mark-ra-states,insert-negate-ra-state-pass,finalize-functions will only run those three and not the rest.

Sorry, my earlier `opt` analogy wasn't the most precise. What I want to test here is that `mark-ra-states` runs, then I treat the transformations between that and `insert-negate-ra-state-pass` as a "black box". So I need these transformations, to introduce new instructions.

What I wanted to simplify with is to skip reverse-engineering all passes in their current state, so I can make them generate new instructions in the exact way I need them to test these changes. Instead, these gtests "pretend" that a certain transformation happened, introducing new instructions in all sort of configurations.

Naturally, this _could mean_ that the changes here are covering cases that would not exist in any real code, but it looks logical to make this part as robust as I can, without looking for specific assembly inputs for each possible case. 

WDYT?

(cc: @paschalis-mpeis, @peterwaller-arm )

bgergely0 commented:
> However, in testing, you want to cover certain scenarios that depend on one or more passes that run between buildCFG and InsertNegateRAState , and these passes are (understandably) considered unstable. Then, the function state, that you are interested in, is impossible to recreate in assembly

Yes, that's it. 

> If that's the case, I think you got the right approach for testing.

All right, thanks!

bgergely0 committed:
[BOLT] Simplify

As fully unknown blocks (stubs) don't have correct unwind info, guessing
their RA state is also futile. Remove code implementing this.

bgergely0 committed:
[BOLT] Fix format && style

bgergely0 committed:
[BOLT] Review

- don't use auto instead of std::optional
- add explicit .has_value() to checking for nullopt (as the
  optional contains a bool)
- fix comment

bgergely0 committed:
[BOLT] Fixes

bgergely0 committed:
[BOLT] Update doc

bgergely0 committed:
[BOLT] Fix: skip unneeded iterations

bgergely0 committed:
[BOLT] Update InsertNegateRAStatePass.h

- remove unused declarations
- update comments

peterwaller-arm commented on bolt/lib/Passes/InsertNegateRAStatePass.cpp:
```diff
@@ -104,32 +118,122 @@ void InsertNegateRAState::coverFunctionFragmentStart(BinaryFunction &BF,
-      PrevInst = Inst;
+      auto Last = BB->getLastNonPseudo();
+      if (Last != BB->rend())
+        PrevInst = *Last;
```
Can we have basic blocks which only contain pseudos? (you've got a check against `rend()`, for example). If this happened, `PrevInst` would be default-initialised, `PrevRAState = BC.MIB->getRAState(PrevInst);` would be `null` and then we could reach the `llvm_unreachable` above? I guess this would require an 'unknown instructions' block to follow a block which has been drained, which perhaps we don't know if this happens in practice. On the other hand, is the current behaviour good enough -- perhaps we can drop the unreachable if we can anyway proceed. 

What are the consequences of hitting the unreachable? Is it easy to make safe?

bgergely0 replied:
Added a test with an empty BB, and removed the `unreachable`, as that seemed the easiest way to change the code.

bgergely0 committed:
Fix for cases where empty BBs are preceding unknown BBs

also added regression test for this.

peterwaller-arm reviewed (approved):
LGTM.

paschalis-mpeis commented on bolt/unittests/Passes/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+  DebugInfoDWARF
+  Object
+  MC
+  ${LLVM_TARGETS_TO_BUILD}
```
```suggestion
  ${BOLT_TARGETS_TO_BUILD}
```

paschalis-mpeis commented on bolt/unittests/Passes/InsertNegateRAState.cpp:
```diff
@@ -0,0 +1,325 @@
+
+  auto CFILoc = findCFIOffsets(*BF);
+  EXPECT_EQ(CFILoc.size(), 1u);
+  EXPECT_EQ(CFILoc[0], 4);
```
Thanks for the extra efforts on unit tests! I think we have an additional opportunity here with little extra work: we could check that the state of `Unknown` and `Unknown1` matches what we would expect. This probably applies to the next test as well.

paschalis-mpeis reviewed (commented):
Hey Gergely,

Thanks for the refactoring. I think it looks good!
I've left a couple of suggestions below.

bgergely0 committed:
Review fixes

peterwaller-arm commented on bolt/unittests/Passes/InsertNegateRAState.cpp:
```diff
@@ -0,0 +1,335 @@
+  Error E = PassManager->runPasses();
+  EXPECT_FALSE(E);
+
+  BF->dump();
```
Intended BF->dump?

paschalis-mpeis reviewed (approved):
Thanks Gergely, looks good. Feel free to merge soon.
